cli/tests/stake.rs

```diff
@@ -542,6 +548,41 @@ fn test_stake_authorize() {
     };
     assert_eq!(current_authority, offline_authority_pubkey);
 
+    // Offline assignment of new nonced stake authority
+    let nonced_authority = Keypair::new();
+    let nonced_authority_pubkey = nonced_authority.pubkey();
+    let (_nonced_authority_file, mut tmp_file) = make_tmp_file();
+    write_keypair(&nonced_authority, tmp_file.as_file_mut()).unwrap();
+    config.command = CliCommand::StakeAuthorize {
+        stake_account_pubkey,
+        new_authorized_pubkey: nonced_authority_pubkey,
+        stake_authorize: StakeAuthorize::Staker,
+        authority: Some(read_keypair_file(&offline_authority_file).unwrap().into()),
+        sign_only: true,
+        signers: None,
+        blockhash: None,
+    };
+    let sign_reply = process_command(&config).unwrap();
+    let (blockhash, signers) = parse_sign_only_reply_string(&sign_reply);
+    config.command = CliCommand::StakeAuthorize {
+        stake_account_pubkey,
+        new_authorized_pubkey: nonced_authority_pubkey,
+        stake_authorize: StakeAuthorize::Staker,
+        // We need to be able to specify the authority by pubkey/sig pair here
```

Comment:
@rob-solana I'm thinking the best option here is to build a wrapper implementing `KeypairUtil` that resolves the command line arg to either a `Keypair`, or maps from a supplied `Pubkey` supplied by `--*-authority` to a `(Pubkey, Signature)` tuple supplied with the `--signer` arg. `sign_message()` would then do a normal signing in the `Keypair` case and in the `(Pubkey, Signature)` case verify+return the signature.  Caveat being that `sign_message()` is currently infallible, so would probably need to panic in case of verify failing.

Something like
```
enum SigningAuthority {
    Online(Keypair),
    Offline(Pubkey, Signature),
}

impl KeypairUtil for SigningAuthority {
    ...
    fn sign_message(&self, msg: &[u8]) {
        match self {
            SigningAuthority::Online(keypair) => keypair.sign_message(msg),
            SigningAuthority::Offline(pubkey, signature) => {
                if signature.verify(pubkey, msg) {
                     signature
                } else {
                     panic!("Signature verification failed!");
                }
            }
        }
    }
}

Reply:
neat!

Reply:
Only thing: last time I tried to make a slice of KeypairUtils from 2 instances of different types, I failed (Rust newb).

Reply:
Fun!  I'll definitely be prototyping outside the codebase.  It's food for another PR in any case

Reply:
> Only thing: last time I tried to make a slice of KeypairUtils from 2 instances of different types, I failed (Rust newb).

Checkout #8318 :wink: